# Parlament Speaker Recogniction Model
This notebook should predict what speaker said certain things based on the Austrian parlament speaches. 

This small project is based on the Dataset assembled by Mario Zechner (https://marioslab.io/)

In [1]:
# DATA FROM SPEAKER
import pandas as pd

# Either use direct download or load locally https://woswormeileistung.marioslab.io/data/persons.json
#speaker = pd.read_json("https://woswormeileistung.marioslab.io/data/persons.json")
speaker = pd.read_json("persons.json")
print(speaker.shape)
speaker.head()

(681, 6)


id                        name      parties  \
0   3612              Sonja Ablinger        [SPÖ]   
1  14854  Dipl.-Ing. Elke Achleitner   [BZÖ, FPÖ]   
2  83119           Mag. Nikolaus Alm  [LIF, NEOS]   
3  51879   Mag. Hannes Amesbauer, BA        [FPÖ]   
4   2819            Werner Amon, MBA        [ÖVP]   

                                        periods  \
0                             [XX, XXIII, XXIV]   
1                                        [XXII]   
2                                         [XXV]   
3                                 [XXVI, XXVII]   
4  [XIX, XX, XXI, XXII, XXIII, XXIV, XXV, XXVI]   

                                    url  \
0   https://parlament.gv.at/person/3612   
1  https://parlament.gv.at/person/14854   
2  https://parlament.gv.at/person/83119   
3  https://parlament.gv.at/person/51879   
4   https://parlament.gv.at/person/2819   

                                            imageUrl  
0  https://parlament.gv.at/dokument/bild/34898/34...  
1  https://parlament.gv.at/dokument/bild/21013/21...  
2  https://parlament.gv.at/dokument/bild/43808/43...  
3  https://parlament.gv.at/dokument/bild/200696/2...  
4  https://parlament.gv.at/dokument/bild/90356/90...

As one can see there are about 681 Speakers from the austrian parlament, starting from 2002

In [2]:
# DATA FROM SESSIONS 
# BE AWARA: sessions data has 440 mb, so it is only imported locally 
# https://woswormeileistung.marioslab.io/data/sessions.json

sessions = pd.read_json("sessions.json")

print(sessions.shape)
sessions.head()

(1001, 7)


url period  sessionNumber  \
0  https://parlament.gv.at/gegenstand/XXVII/NRSIT...  XXVII            251   
1  https://parlament.gv.at/gegenstand/XXVII/NRSIT...  XXVII            250   
2  https://parlament.gv.at/gegenstand/XXVII/NRSIT...  XXVII            249   
3  https://parlament.gv.at/gegenstand/XXVII/NRSIT...  XXVII            248   
4  https://parlament.gv.at/gegenstand/XXVII/NRSIT...  XXVII            247   

                sessionLabel       date protocolUrls sections  
0  251. Sitzung (251/NRSITZ) 2024-01-31           []       []  
1  250. Sitzung (250/NRSITZ) 2024-01-31           []       []  
2  249. Sitzung (249/NRSITZ) 2024-01-31           []       []  
3  248. Sitzung (248/NRSITZ) 2023-12-15           []       []  
4  247. Sitzung (247/NRSITZ) 2023-12-15           []       []

In [3]:
sessions.sections.head()

0    []
1    []
2    []
3    []
4    []
Name: sections, dtype: object

In [4]:
sessions.sections.drop_duplicates()

0                                                      []
35      [{'speaker': '88386', 'text': 'Ich darf die 21...
36      [{'speaker': '88386', 'text': 'Meine sehr geeh...
39      [{'speaker': '88386', 'text': 'Ich eröffne die...
40      [{'speaker': '88386', 'text': 'Meine sehr geeh...
                              ...                        
996     [{'speaker': '799', 'text': 'Ich eröffne die 5...
997     [{'speaker': '799', 'text': 'Die Sitzung ist e...
998     [{'speaker': '799', 'text': 'Meine Damen und H...
999     [{'speaker': '799', 'text': 'Die Sitzung ist e...
1000    [{'speaker': '334', 'text': 'Meine Damen und H...
Name: sections, Length: 955, dtype: object

In [5]:
sessions.sections.describe()

count     1001
unique     955
top         []
freq        47
Name: sections, dtype: object

In [6]:
session_na = [1 if sessions["sections"][i] == [] else 0 for i in range(len(sessions))]
print("Absolute number of na: "+ str(sum(session_na)) + " || relative number of na: "+ str(sum(session_na)/len(sessions)))

Absolute number of na: 47 || relative number of na: 0.04695304695304695


In [7]:
# the entries in "speaker" arer list of dicts; therefore loop over the list to extract the dicts
text_dict = [j  for sections in sessions["sections"] for j in sections] 

# make a pandas df from the extracted dict
text = pd.DataFrame.from_dict(text_dict, orient="columns")
text["input"] = "SPEAKER: "+text["speaker"]+"; TEXT: "+text["text"]
print(text.shape) # 173498 
text.head()

(173498, 5)


speaker                                               text  \
0   88386  Ich darf die 216. Sitzung des Nationalrates um...   
1   88386  Hinsichtlich der eingelangten Verhandlungsgege...   
2   88386  Außerdem weise ich\n\ndie Anträge 3425/A, 3426...   
3   88386  Weiters darf ich mitteilen, dass folgende Fris...   
4   88386  Wir kommen nunmehr zur Abstimmung über den Ant...   

                                            callouts  \
0  [{'caller': '2343', 'text': 'Ich bin noch nich...   
1                                                 []   
2                                                 []   
3                                                 []   
4                                                 []   

                                               links  \
0                                                 []   
1  [{'label': '14141/AB', 'url': 'https://parlame...   
2  [{'label': '3425/A', 'url': 'https://parlament...   
3  [{'label': '1774 der Beilagen', 'url': 'https:...   
4  [{'label': '1774 der
Beilagen', 'url': 'https:...   

                                               input  
0  SPEAKER: 88386; TEXT: Ich darf die 216. Sitzun...  
1  SPEAKER: 88386; TEXT: Hinsichtlich der eingela...  
2  SPEAKER: 88386; TEXT: Außerdem weise ich\n\ndi...  
3  SPEAKER: 88386; TEXT: Weiters darf ich mitteil...  
4  SPEAKER: 88386; TEXT: Wir kommen nunmehr zur A...

In [8]:
#for i in range(10): 
#    print(text["callouts"][i] == [])
tryout = pd.DataFrame([text["callouts"][i] == [] for i in range(10)])
tryout.describe()

0
count     10
unique     2
top     True
freq       9

# Try using Hugging Face transformers library

In [9]:
from datasets import Dataset,DatasetDict

ds = Dataset.from_pandas(text[:100])
ds

Dataset({
    features: ['speaker', 'text', 'callouts', 'links', 'input'],
    num_rows: 173498
})

In [10]:
from transformers import AutoModelForSequenceClassification,AutoTokenizer
model_nm = 'microsoft/deberta-v3-small'
t#okz = AutoTokenizer.from_pretrained(model_nm)

def tok_func(x): return tokz(x["input"])
#tok_ds = ds.map(tok_func, batched=True)
#tok_ds

/home/georg/miniforge3/lib/python3.10/site-packages/transformers/convert_slow_tokenizer.py:515: UserWarning: The sentencepiece tokenizer that you are converting to a fast tokenizer uses the byte fallback option which is not implemented in the fast tokenizers. In practice this means that the fast version of the tokenizer can produce unknown tokens whereas the sentencepiece version would have converted these unknown tokens into a sequence of byte tokens matching the original piece of text.
  warnings.warn(


In [11]:
# MY JOURNEY WITH THE HUGGING FACE LIBRARY STOPS HERE BECAUSE OF AN ERROR

#from transformers import TrainingArguments,Trainer

#args = TrainingArguments('outputs', 
#                         learning_rate=8e-5, 
#                         warmup_ratio=0.1, 
#                         lr_scheduler_type='cosine', 
#                         fp16=False,
#                         evaluation_strategy="epoch", 
#                         per_device_train_batch_size=128, 
#                         per_device_eval_batch_size=128*2,
#                         num_train_epochs=4, 
#                         weight_decay=0.01, 
#                         report_to='none')

In [12]:
#model = AutoModelForSequenceClassification.from_pretrained(model_nm, num_labels=1)
#trainer = Trainer(model, args, train_dataset=dds['train'], eval_dataset=dds['test'],tokenizer=tokz, compute_metrics=corr_d)

# Try the speaker classification with fast.ai library

In [13]:
from fastai.text.core import tokenize_df
train_df = text[:100]
text_tok = tokenize_df(df=train_df,text_cols="text")

In [14]:
text_tok[0]

speaker  \
0    88386   
1    88386   
2    88386   
3    88386   
4    88386   
..     ...   
95   83129   
96   88386   
97   18140   
98   88386   
99    2242   

                                                                                                                                                 callouts  \
0   [{'caller': '2343', 'text': 'Ich bin noch nicht fertig, Herr Präsident!'}, {'text': 'Beifall bei SPÖ und Grünen.'}, {'text': 'In Richtung Galerie:'}]   
1                                                                                                                                                      []   
2                                                                                                                                                      []   
3                                                                                                                                                      []   
4                                                                                                                                                      []   
..                                                                                                                                                    ...   
95                                                                                                                                                     []   
96                                                                                                                                                     []   
97                                                                                                                                                     []   
98                                                                                                                                                     []   
99                                                                                                                                                     []   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      links  \
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        []   
1   [{'label': '14141/AB', 'url': 'https://parlament.gv.at/pls/portal/hi.link?pfad=/PAKT/VHG/XXVII/AB/AB_14141/index.shtml'}, {'label': '14148/AB', 'url': 'https://parlament.gv.at/pls/portal/hi.link?pfad=/PAKT/VHG/XXVII/AB/AB_14148/index.shtml'}, {'label': '3415/A', 'url': 'https://parlament.gv.at/PAKT/VHG/XXVII/A/A_03415/index.shtml'}, {'label': '3410/A', 'url': 'https://parlament.gv.at/PAKT/VHG/XXVII/A/A_03410/index.shtml'}, {'label': '3422/A(E)', 'url': 'https://parlament.gv.at/PAKT/VHG/XXVII/A/A_03422/index.shtml'}, {'label': '3408/A(E)', 'url': 'https://parlament.gv.at/PAKT/VHG/XXVII/A/A_...   
2                                                                                                                               

In [15]:
from fastai.text.data import TextDataLoaders
dl = TextDataLoaders.from_df(train_df, valid_pct=0.2, seed=None,text_col="text", label_col="speaker")
dl.show_batch(max_n=2)

In [39]:
from fastai.metrics import accuracy,Perplexity
from fastai.text.learner import text_classifier_learner
from fastai.text.models import AWD_LSTM
learn = text_classifier_learner(dl, AWD_LSTM, metrics=[accuracy, Perplexity()])

In [41]:
from fastai.text.all import *
learn.fit_one_cycle(1, 2e-2)

IOStream.flush timed out


KeyError: Caught KeyError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/transforms.py", line 261, in encodes
    return TensorCategory(self.vocab.o2i[o])
KeyError: '83121'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/georg/miniforge3/lib/python3.10/site-packages/torch/utils/data/_utils/worker.py", line 308, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/torch/utils/data/_utils/fetch.py", line 41, in fetch
    data = next(self.dataset_iter)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/load.py", line 138, in create_batches
    yield from map(self.do_batch, self.chunkify(res))
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/basics.py", line 230, in chunked
    res = list(itertools.islice(it, chunk_sz))
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/load.py", line 168, in do_item
    try: return self.after_item(self.create_item(s))
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/load.py", line 175, in create_item
    if self.indexed: return self.dataset[s or 0]
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/core.py", line 447, in __getitem__
    res = tuple([tl[it] for tl in self.tls])
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/core.py", line 447, in <listcomp>
    res = tuple([tl[it] for tl in self.tls])
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/core.py", line 406, in __getitem__
    return self._after_item(res) if is_indexer(idx) else res.map(self._after_item)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/core.py", line 366, in _after_item
    def _after_item(self, o): return self.tfms(o)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/transform.py", line 208, in __call__
    def __call__(self, o): return compose_tfms(o, tfms=self.fs, split_idx=self.split_idx)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/transform.py", line 158, in compose_tfms
    x = f(x, **kwargs)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/transform.py", line 81, in __call__
    def __call__(self, x, **kwargs): return self._call('encodes', x, **kwargs)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/transform.py", line 91, in _call
    return self._do_call(getattr(self, fn), x, **kwargs)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/transform.py", line 97, in _do_call
    return retain_type(f(x, **kwargs), x, ret)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastcore/dispatch.py", line 120, in __call__
    return f(*args, **kwargs)
  File "/home/georg/miniforge3/lib/python3.10/site-packages/fastai/data/transforms.py", line 263, in encodes
    raise KeyError(f"Label '{o}' was not included in the training dataset") from e
KeyError: "Label '83121' was not included in the training dataset"


In [ ]:
learn.predict("Ich bin Wolfgang Sobotka.")

In [44]:
#speaker["id"] == 18140
speaker.loc[speaker["id"]==18140]

id                    name parties  periods  \
600  18140  Mag. Dr. Martin Kocher      []  [XXVII]   

                                      url  \
600  https://parlament.gv.at/person/18140   

                                                          imageUrl  
600  https://parlament.gv.at/dokument/bild/200504/20050490_384.jpg

In [47]:
len(train_df["speaker"].unique())

17